File: app/src/main/java/cl/figonzal/evaluatool/adapter/EvaluaAdapter.java
/*--------------------------------------------------------------
                                                               -
 This file is subject to the terms and conditions defined in   -
 file 'LICENSE', which is part of this source code package.    -
                                                               -
 Autor: Felipe González                                        -
 Email: felipe.gonzalezalarcon94@gmail.com                     -
                                                               -
 Copyright (c) 2020.                                           -
                                                               -
 Last modified 09-03-20 16:51                                  -
 --------------------------------------------------------------*/

package cl.figonzal.evaluatool.adapter;


import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

import cl.figonzal.evaluatool.R;
import cl.figonzal.evaluatool.modelo.Evalua;
import io.github.luizgrp.sectionedrecyclerviewadapter.Section;
import io.github.luizgrp.sectionedrecyclerviewadapter.SectionParameters;

public class EvaluaAdapter extends Section {

    private final String nombre_header;
    private final List<Evalua> subItems;
    private final ClickListener clickListener;

    private boolean expanded = false;

    public EvaluaAdapter(@NonNull String nombre_header, @NonNull List<Evalua> list, ClickListener listener) {
        super(
                SectionParameters.builder()
                        .itemResourceId(R.layout.expandable_list_item)
                        .headerResourceId(R.layout.expandable_list_group)
                        .build()
        );

        this.nombre_header = nombre_header;
        this.subItems = list;
        this.clickListener = listener;
    }

    @Override
    public int getContentItemsTotal() {
        return expanded ? subItems.size() : 0;
    }

    @Override
    public RecyclerView.ViewHolder getItemViewHolder(View view) {
        return new ItemViewHolder(view);
    }

    @Override
    public void onBindItemViewHolder(final RecyclerView.ViewHolder holder, int position) {

        final ItemViewHolder itemViewHolder = (ItemViewHolder) holder;

        Evalua evalua = subItems.get(position);

        itemViewHolder.tv_item_title.setText(evalua.getNombre());
        itemViewHolder.itemView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                clickListener.onItemRootViewClicked(nombre_header, itemViewHolder.getAdapterPosition());
            }
        });

    }

    @Override
    public void onBindHeaderViewHolder(final RecyclerView.ViewHolder holder) {
        HeaderViewHolder headerViewHolder = (HeaderViewHolder) holder;

        headerViewHolder.tv_header_title.setText(nombre_header);
        headerViewHolder.iv_arrow.setImageResource(
                expanded ? R.drawable.ic_keyboard_arrow_up_black_24dp : R.drawable.ic_keyboard_arrow_down_black_24dp
        );

        headerViewHolder.rootView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                clickListener.onHeaderRootViewClicked(EvaluaAdapter.this);
            }
        });

    }

    @Override
    public RecyclerView.ViewHolder getHeaderViewHolder(View view) {
        return new HeaderViewHolder(view);
    }

    public boolean isExpanded() {
        return expanded;
    }

    public void setExpanded(final boolean expanded) {
        this.expanded = expanded;
    }

    public interface ClickListener {

        void onHeaderRootViewClicked(@NonNull final EvaluaAdapter section);

        void onItemRootViewClicked(@NonNull final String sectionTitle, final int itemAdapterPosition);
    }

    /**
     * Clase holder para headers de lista expandida
     */
    static class HeaderViewHolder extends RecyclerView.ViewHolder {

        private final View rootView;
        private final TextView tv_header_title;
        private final ImageView iv_arrow;

        HeaderViewHolder(@NonNull View view) {
            super(view);

            rootView = view;
            tv_header_title = view.findViewById(R.id.tv_group);
            iv_arrow = view.findViewById(R.id.iv_arrow);
        }
    }

    /**
     * Clase holder para items de lista expandida
     */
    static class ItemViewHolder extends RecyclerView.ViewHolder {

        private final TextView tv_item_title;

        ItemViewHolder(@NonNull View itemView) {
            super(itemView);
            tv_item_title = itemView.findViewById(R.id.tv_child);
        }
    }
}

File: app/src/main/java/cl/figonzal/evaluatool/adapter/EvaluaAdapter.kt
/*

 This file is subject to the terms and conditions defined in
 file 'LICENSE', which is part of this source code package

 Autor: Felipe González
 Email: felipe.gonzalezalarcon94@gmail.com

 Copyright (c) 2021

 Last modified 18-04-21 18:20
 */
package cl.figonzal.evaluatool.adapter

import android.view.View
import androidx.recyclerview.widget.RecyclerView
import cl.figonzal.evaluatool.R
import cl.figonzal.evaluatool.databinding.ExpandableItemGroupListBinding
import cl.figonzal.evaluatool.databinding.ExpandableItemListBinding
import cl.figonzal.evaluatool.modelo.Evalua
import io.github.luizgrp.sectionedrecyclerviewadapter.Section
import io.github.luizgrp.sectionedrecyclerviewadapter.SectionParameters

/**
 * Adapter that is used
 */
class EvaluaAdapter(private val headerName: String, private val subItemsList: List<Evalua>, private val clickListener: ClickListener) : Section(
        SectionParameters.builder()
                .itemResourceId(R.layout.expandable_item_list)
                .headerResourceId(R.layout.expandable_item_group_list)
                .build()
) {
    var isExpanded = false

    override fun getContentItemsTotal(): Int = if (isExpanded) subItemsList.size else 0

    override fun getItemViewHolder(view: View): RecyclerView.ViewHolder = ItemViewHolder(view)
    override fun getHeaderViewHolder(view: View): RecyclerView.ViewHolder = HeaderViewHolder(view)

    override fun onBindItemViewHolder(holder: RecyclerView.ViewHolder, position: Int) {

        //Casting
        val itemViewHolder = holder as ItemViewHolder

        itemViewHolder.bind(subItemsList[position], this)
    }

    override fun onBindHeaderViewHolder(holder: RecyclerView.ViewHolder) {

        //Casting
        val headerViewHolder = holder as HeaderViewHolder

        headerViewHolder.bind(this)
    }

    interface ClickListener {
        fun onHeaderRootViewClicked(section: EvaluaAdapter)
        fun onItemRootViewClicked(sectionTitle: String, itemAdapterPosition: Int)
    }

    /**
     * Clase holder para headers de lista expandida
     * ViewHolder for headers of Evalua's list
     *
     * @param rootView Parent view
     */
    internal class HeaderViewHolder(rootView: View) : RecyclerView.ViewHolder(rootView) {

        val binding = ExpandableItemGroupListBinding.bind(rootView)

        /**
         * Binding resources of Header view
         *
         * @param evaluaAdapter PassAdapter to provide all the required information
         */
        fun bind(evaluaAdapter: EvaluaAdapter) {

            with(binding, {

                tvGroup.text = evaluaAdapter.headerName
                ivArrow.setImageResource(
                        when {
                            evaluaAdapter.isExpanded -> R.drawable.ic_keyboard_arrow_up_black_24dp
                            else -> R.drawable.ic_keyboard_arrow_down_black_24dp
                        }
                )
                root.setOnClickListener {
                    evaluaAdapter.clickListener.onHeaderRootViewClicked(evaluaAdapter)
                }
            })
        }
    }

    /**
     * Clase holder para items de lista expandida
     */
    internal class ItemViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        val binding = ExpandableItemListBinding.bind(itemView)

        /**
         * Binding resources for items list
         *
         * @param subItem SubItem binded
         * @param evaluaAdapter PassAdapter to provide all the required information
         */
        fun bind(subItem: Evalua, evaluaAdapter: EvaluaAdapter) {

            with(binding, {
                tvChild.text = subItem.nombre
                itemView.setOnClickListener { evaluaAdapter.clickListener.onItemRootViewClicked(evaluaAdapter.headerName, adapterPosition) }
            })
        }
    }
}